### DISCUSSION TITLE
JIT function frozen... minimization failure

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#10648] [Q&A]
Hi, 
Trying to make a helper function to allow for jit-method in class, I face a problem in the following use-case of minimization when I would like to minimize the number of recompilation of the `_test` function. For the sake of completeness I use the `jaxopt.ScipyBoundedMinimize` minimizer which cannot be jitted. 

Below is the snippet where for "res 1" I instantiate a new `A` object at each time which yields a recompilation at each time and gives the right expected solution (minimum is equal to `a`). For the "res 2" case, a single `A` object is instantiate once for all, and it yields a single compilation of `_test`BUT the results are only correct for the first `a`value. 

I do not yet manage to reconcile 1) right solution and 2) a single compilation. Have you a comment/idea? Thanks in advance. 

```python
import jax
import jax.numpy as jnp
from jax import vmap, jit
jax.config.update("jax_enable_x64", True)

import jaxopt

@jit
def _test(x, a):
    print("compile....")
    return (x-a)**2

class A():
    def __init__(self):
        
        tol = 1e-3
        method = 'L-BFGS-B'
        options = {'disp':False,'ftol':tol, 'gtol':tol, 'maxiter':600}
        self.jscMin=jaxopt.ScipyBoundedMinimize(fun=self.test,
                                           method=method, 
                                           tol=tol,
                                           options=options)
    
        self.Init()
    
    def Init(self):
        self.a = None
        
    def set_a(self, val):
        self.a = val
    
    def test(self, x):
        return _test(x, self.a)

    def optimize(self, init_val, low_bnd, high_bnd):
        res = self.jscMin.run(init_val,bounds=(low_bnd,high_bnd))
        return res
    
        
    

print(">>> res1")
for a in [2., -5., 10.]:
    obja = A()
    obja.set_a(a)
    print("a: ",obja.a)
    res = obja.optimize(jnp.array(0.),-10.,10.)
    print(res)

    
print(">>> res2")
obja = A()
for a in [2., -5., 10.]:
    obja.set_a(a)
    print("a: ",obja.a)
    res = obja.optimize(jnp.array(0.),-10.,10.)
    print(res)
```
leading to
```
>>> res1
a:  2.0
compile....
OptStep(params=DeviceArray(2., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  -5.0
compile....
OptStep(params=DeviceArray(-5., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  10.0
compile....
OptStep(params=DeviceArray(10., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=1))


>>> res2
a:  2.0
compile....
OptStep(params=DeviceArray(2., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  -5.0
OptStep(params=DeviceArray(2., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  10.0
OptStep(params=DeviceArray(2., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
```

**Comment by jecampagne** [comment]
Ha, I got it

In fact, `jaxopt.ScipyBoundedMinimize` implicitly trigger a compilation to minimize. So in my original snippet, this is the `test(self,x)`. But then, comes the `self` problem which is frozen, so the `set_a` is completely useless as the first `a`value known at compilation is `2`.

In the code below, the function argument of `jaxopt.ScipyBoundedMinimize` is directly the helper function, and I adjust the argument of `self.jscMin.run` in `optimize` to set the proper `a`-value

```python
@jit
def _test(x, a):
    print("compile....")
    return (x-a)**2

class A():
    def __init__(self):
        
        tol = 1e-3
        method = 'L-BFGS-B'
        options = {'disp':False,'ftol':tol, 'gtol':tol, 'maxiter':600}
        self.jscMin=jaxopt.ScipyBoundedMinimize(fun=_test,
                                           method=method, 
                                           tol=tol,
                                           options=options)
    
        self.Init()
    
    def Init(self):
        self.a = None
        
    def set_a(self, val):
        self.a = val
    
#    def test(self, x):
#        return _test(x, self.a)

    def optimize(self, init_val, low_bnd, high_bnd):
        res = self.jscMin.run(init_val,bounds=(low_bnd,high_bnd), a=self.a)
        return res
 
```
It yields
```
>>> res2
a:  2.0
compile....
OptStep(params=DeviceArray(2., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  -5.0
OptStep(params=DeviceArray(-5., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=2))
a:  10.0
OptStep(params=DeviceArray(10., dtype=float64), state=ScipyMinimizeInfo(fun_val=DeviceArray(0., dtype=float64, weak_type=True), success=True, status=0, iter_num=1))
```

=> 1 single compilation and all minimisations are correct !. It is a good solution or do you see some possible failures?

**Reply by jakevdp** [reply]
Yes, this looks like a decent solution – but your code still relies on in-place object mutation (the `set_a` method) and doesn't define how the object should be treated in terms of the JIT cache, so, for example, if you tried to JIT a function that called `self.optimize` you might still obtain incorrect results because updates to `self.a` would not lead to cache misses. The FAQ addresses this: https://jax.readthedocs.io/en/latest/faq.html#how-to-use-jit-with-methods In particular, if you're using this approach you should probably register your function as a PyTree (option 3) to have a safer implementation. For what it's worth, ~I believe that if you had done this from the beginning, your initial implementation would have been fine as well, though I'll admit I've not deeply explored the mechanics of JIT compilation of bound methods interacting with object mutation~. *Edit: I tried it and it seems that jit-compiling bound methods will lead to executions that don't respect in-place updates of object attributes: so probably your best bet is to not do this.*

Even better would be to stop relying on complicated object mutation and self-reference, and just write plain functional code in which the outputs of a function only depend on the inputs – that's the paradigm that JAX transforms like `jit` are designed for (see https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions), and any time you're trying to bypass that with clever cacheing tricks, you're bound to face problems.

For example, here's a version of your code that's much more obviously correct, and IMHO much easier to understand at a first glance than the class-based obfuscation of the original:
```python
@jit
def _test(x, a):
    return (x-a)**2

tol = 1E-5
jacMin = jaxopt.ScipyBoundedMinimize(fun=_test,
                                     method='L-BFGS-B', 
                                     tol=tol,
                                     options={'disp':False,'ftol':tol, 'gtol':tol, 'maxiter':600})
for a in [2., -5., 10.]:
    print("a: ", a)
    res = jacMin.run(jnp.array(0.), bounds=(-10.,10.), a=a)
    print(res)
```

**Reply by jecampagne** [reply]
Hum. It is my bkgd on C++ for 25yrs but it seems that C-coding style is may be a good JAX thinking?

**Reply by jakevdp** [reply]
I think object-oriented approaches are fine, so long as you teach JAX to handle the objects properly (e.g. pytree registration) and as long as your functions are pure (i.e. in most cases don't rely on in-place object mutation). The two FAQ links above are the most relevant resources on these, I think.

**Reply by jecampagne** [reply]
When you write "In particular, if you're using this approach you should probably register `your function` as a PyTree (option 3) to have a safer implementation.", which function do you have in mind?

**Reply by jakevdp** [reply]
Sorry for the typo: I meant register your class.

**Reply by jecampagne** [reply]
It could be nice to have the skeleton of "A" well adapted in the context of PyTree (at least as exercice). Can you help?
Thanks
(nb. many of the exchanges are not visible and I only receive mails but lack to see them in this thread?!)

**Reply by PhilipVinc** [reply]
An easy way to do this is to use `@flax.struct.dataclass` decorator, and declare the data class fields as nodes or not accordingly...
Something like (there might be typos)

The only thing you should keep in mind is that things that should retrigger recompilation if they change should be tagged with `pytree_node=False`, the others will get the default `pytre_node=True`.
`Pytree_node=False` will lead to similar behaviour to declaring that field a `static_arg` of jax.jit.

```python
from typing import Any, Callable

@flax.struct.dataclass
class A():
   fun: Callable = flax.struct.field(pytree_node=False)
   args: Any 
   tol : float = 1e-3
   method : str = flax.struct.field(pytree_node=False, default= 'L-BFGS-B')
   disp: str = flax.struct.field(pytree_node=False, default=False)
   ftol: float = 1e-5
   gtol : float = 1e-5
   maxiter : int = 600 
   
   def __post_init__(self):
        jscMin=jaxopt.ScipyBoundedMinimize(fun=seelf.fun,
                                           method=self.method, 
                                           tol=self.tol,
                                           maxiter=self.maxiter)
        object.__setattr__(self, 'jscMin', jscMin)
    

    def optimize(self, init_val, low_bnd, high_bnd):
        return self.jscMin.run(init_val,bounds=(low_bnd,high_bnd), *self.args)
```

**Reply by jecampagne** [reply]
Nice imput @PhilipVinc  I will try right now.

**Reply by PhilipVinc** [reply]
I got curious and actually tried running the code. Here is a version that runs https://gist.github.com/PhilipVinc/47c762189aab974c9a719a243d519f9a 

there's a commented out `jax.jit` to jit the full `optimise` method, but it does not work.
I believe it does not because `jaxopt.ScipyBoundedMinimize` is trying to convert to numpy the `bounds` arguments. 
Basically, it's not jax.compatible.

Try using another solver?

**Reply by PhilipVinc** [reply]
Yep https://github.com/google/jaxopt/issues/214

**Reply by jakevdp** [reply]
I think there is a bug here:
```python
jscMin=jaxopt.ScipyBoundedMinimize(fun=self.fun, ...)
```
You are passing a bound method to the minimizer, which will effectively wrap that bound method in a `jit`. Since the bound method has implicit rather than explicit dependence on `self`, after its first call any mutation of `self` will not lead to a cache miss, so you effectively freeze the contents of `self` as it appears during the first execution.

I stick by my original recommendation: your best bet is to stop trying to work around JAX's functional design, and just write pure/functional code.

**Reply by PhilipVinc** [reply]
Well, that's a frozen data class so you can't mutate it without rerunning the init code...

**Reply by jecampagne** [reply]
Correct and as you probably see it is one of my issue on jaxopt ... *the world is too small*.

Well, this solver was well adapted to my task, I was using "L-BFGS-B" to get optimal 6-hyper-parameters of a Gaussian Process (Rasmussen& William book where there is polynomial feature means and a gaussian prior of the linear coeff.). There might be an other JAX-compatible minimizer well adapted (an suggestion is welcome)

**Reply by jakevdp** [reply]
> Well, that's a frozen data class so you can't mutate it without rerunning the init code...

Ah, thanks, I didn't realize that. Most of the issues in this question arise from the use of mutation, so +💯 to using any structure that makes that illegal 😁

**Reply by PhilipVinc** [reply]
I believe that a PR or just a hack to make `jaxopt.ScipyOptimizer` use a host_callback would be relatively simple  as you only need to edit this line (and 2 lines above) https://github.com/google/jaxopt/blob/85313307346fa2109266b0f43cef05460043f302/jaxopt/_src/scipy_wrappers.py#L287 

The problem is that if you then jit the function `f` to optimise,  then you'd have a python -> jit -> scipyOptimizer -> callback -> jit -> f` and calling back a jittedfunction from a callback leads to deadlocks on GPUs (though it works fine on CPUs, in my experience).

I don't even know if calling jax-jitted code from callbacks is officially supported. I'd wager it isn't but maybe @jakevdp knows?

**Reply by jecampagne** [reply]
Well @PhilipVinc  I think you are more experienced than me to justify/write a PR to jaxopt. My only once-cent was to rise the problem (from a user point of view).

**Reply by jecampagne** [reply]
@PhilipVinc 
I do not know if using `optax` may help for my task as the hyper-parameters search should be bounded for some of them other wise  it will fail...

here a example of optax use throw jaxopt: where nl_lik is a non-linear function which depends on parameters and (Xi,yi) training samples
```python   
opt = optax.adam(0.01)
solver = jaxopt.OptaxSolver(opt=opt, fun=nl_lik, maxiter=600)
init_params = jnp.array([1.,2.])
res = solver.run(init_params, X=xi, y=yi)
```
but the point here is does it can be included in the solution you have proposed in place of ScipyBoudedMinimize.

**Reply by PhilipVinc** [reply]
I guess it can. Just try

**Reply by jecampagne** [reply]
Well I face pb to install [optax (see issue #341)](https://github.com/deepmind/optax/issues/341)

**Reply by jecampagne** [reply]
Well, `optax.adam` is absolutely not adapted :( . I will try [jaxopt.ProjectedGradient](https://jaxopt.github.io/stable/_autosummary/jaxopt.ProjectedGradient.html#jaxopt.ProjectedGradient) with [.projection_box](https://jaxopt.github.io/stable/_autosummary/jaxopt.projection.projection_box.html#jaxopt.projection.projection_box),  but do not know it is JAX-compatible (contrary to `ScipyBoundedMinimize`)

**Reply by jecampagne** [reply]
Well after investigation [jaxopt.ProjectedGradient](https://jaxopt.github.io/stable/_autosummary/jaxopt.ProjectedGradient.html#jaxopt.ProjectedGradient) with [.projection_box](https://jaxopt.github.io/stable/_autosummary/jaxopt.projection.projection_box.html#jaxopt.projection.projection_box) is not adapted too. So I need certainly to find a L-BFGS-B solver with bounded hyper-cube constraint compatible with JAX...

**Reply by PhilipVinc** [reply]
how comes projected gradient is not jax compatible?

**Reply by jecampagne** [reply]
ho, may be my english is not ad equa, jaxopt.ProjectedGradient is certainly JAX compatible, but for my use case it does not manage to get right.

**Reply by jecampagne** [reply]
I just see that `jax.scipy.optimize.minimize` has a BFGS method, (and probably LBFGS is in the pipeline) but does it is possible to impose  "box-bounds" on parameters? Thanks
